File: Commons/Graph/src/main/java/com/greentree/commons/graph/RootTree.java
package com.greentree.commons.graph;

import java.util.ArrayList;

import com.greentree.commons.graph.algorithm.lca.BruteLCAFinder;
import com.greentree.commons.util.iterator.IteratorUtil;


public interface RootTree<V> extends Tree<V> {
	
	default boolean containsInSubTree(Object v, Object parent) {
		for(var e : getChildrenClosure(parent))
			if(v.equals(e))
				return true;
		return false;
	}
	
	default V lca(Object a, Object b) {
		final var finder = new BruteLCAFinder<>(this);
		return finder.lca(a, b);
	}
	
	default V getRoot() {
		final var iter = iterator();
		V p, v;
		p = iter.next();
		do {
			v = p;
			p = getParent(v);
		}while(v != p);
		return v;
	}
	
	default int depth(Object v) {
		int depth = 0;
		V p = null;
		while(true) {
			p = getParent(v);
			if(p == v)
				return depth;
			depth++;
			v = p;
		}
	}
	
	/** @return parent of vertex, or root if vertex is root */
	V getParent(Object v);
	
	default Iterable<? extends V> getChildren(Object v) {
		final var result = new ArrayList<V>();
		for(var e : this) {
			final var p = getParent(e);
			if(p.equals(v))
				result.add(e);
		}
		return result;
	}
	
	@SuppressWarnings("unchecked")
	default Iterable<? extends V> getChildrenClosure(Object v) {
		final var result = new ArrayList<V>();
		result.add((V) v);
		for(var c : getChildren(v)) {
			for(var e : getChildrenClosure(c))
				result.add(e);
		}
		return result;
	}
	
	@Override
	default Iterable<? extends V> getAdjacencyIterable(Object v) {
		return IteratorUtil.union(getChildren(v), IteratorUtil.iterable(getParent(v)));
	}
	
	
}

File: Commons/Graph/src/main/java/com/greentree/commons/graph/RootTree.kt
package com.greentree.commons.graph

import com.greentree.commons.graph.algorithm.lca.BruteLCAFinder

interface RootTree<out V : Any> : Tree<V> {

	fun containsInSubTree(v: @UnsafeVariance V, parent: @UnsafeVariance V): Boolean {
		for(e in getChildrenClosure(parent))
			if(v == e)
				return true
		return false
	}

	fun lca(a: @UnsafeVariance V, b: @UnsafeVariance V): V {
		val finder = BruteLCAFinder(this)
		return finder.lca(a, b)
	}

	val root: V
		get() {
			val iter = iterator()
			var p: V?
			var v: V?
			p = iter.next()
			do {
				v = p
				p = getParent(v)
			} while(v !== p)
			return v!!
		}

	/** @return parent of vertex, or root if vertex is root
	 */
	fun getParent(v: Any?): V

	fun depth(v: Any?): Int {
		var v = v
		var depth = 0
		var p: V? = null
		while(true) {
			p = getParent(v)
			if(p === v) return depth
			depth++
			v = p
		}
	}

	fun getChildrenClosure(v: @UnsafeVariance V): Iterable<V> {
		val result = ArrayList<V>()
		result.add(v)
		for(c in getChildren(v)) {
			for(e in getChildrenClosure(c)) result.add(e)
		}
		return result
	}

	override fun getAdjacencySequence(v: @UnsafeVariance V): Sequence<V> {
		return sequenceOf(getChildren(v), sequenceOf(getParent(v))).flatten()
	}

	fun getChildren(v: @UnsafeVariance V) = sequence {
		for(e in this@RootTree) {
			if(v == getParent(e))
				yield(e)
		}
	}
}
